src/main/cpp/checkpoint/CheckPoint.cpp

```diff
@@ -0,0 +1,681 @@
+
+	if (info->data_size == 0) {
+		result.log_config->output_prefix = "";
+	}
+
+	delete info;
+
+	// travel model
+
+	WriteDSetFile("tmp_pop.csv", "Config/popconfig");
+	std::string popconfig = "tmp_pop.csv";
+	std::string geoconfig = "";
+
+	exist = H5Lexists(m_file, "Config/geoconfig", H5P_DEFAULT);
+	if (exist > 0) {
+		WriteDSetFile("tmp_geoconfig.xml", "Config/geoconfig");
+		geoconfig = "tmp_geoconfig.xml";
+	}
+	std::string household = "";
+	exist = H5Lexists(m_file, "Config/household", H5P_DEFAULT);
+	if (exist > 0) {
+		WriteDSetFile("tmp_household.xml", "Config/household");
+		household = "tmp_household.xml";
+	}
+	auto travelRef = make_shared<multiregion::RegionTravel>(id, popconfig, geoconfig, household);
+	result.travel_model = travelRef;
+
+	return result;
+}
+
+void CheckPoint::SplitCheckPoint(unsigned int groupnum, const std::string& filename)
+{
+	std::stringstream ss;
+	ss << "Simulation " << groupnum;
+	std::string groupname = ss.str();
+
+	htri_t exist = H5Lexists(m_file, groupname.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		FATAL_ERROR("Non-existing group");
+	}
+
+	hid_t group = H5Gopen2(m_file, groupname.c_str(), H5P_DEFAULT);
+
+	hid_t f = H5Fcreate(filename.c_str(), H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);
+
+	auto op_func = [&f](hid_t loc_id, const char* name, const H5L_info_t* info, void* operator_data) {
+		H5Ocopy(loc_id, name, f, name, H5P_DEFAULT, H5P_DEFAULT);
+		return 0;
+	};
+	auto temp = [](hid_t loc_id, const char* name, const H5L_info_t* info, void* operator_data) {
+		(*static_cast<decltype(op_func)*>(operator_data))(loc_id, name, info, operator_data);
+		return 0;
+	};
+
+	H5Literate(group, H5_INDEX_NAME, H5_ITER_NATIVE, nullptr, temp, &op_func);
+
+	H5Fclose(f);
+	H5Gclose(group);
+}
+
+Calendar CheckPoint::LoadCalendar(boost::gregorian::date date)
+{
+	htri_t exist = H5Lexists(m_file, "Config", H5P_DEFAULT);
+	if (exist <= 0) {
+		FATAL_ERROR("Invalid file");
+	}
+
+	WriteDSetFile("tmp_holidays.json", "Config/holidays");
+
+	Calendar result;
+	result.Initialize(date, "tmp_holidays.json");
+	boost::filesystem::remove("tmp_holidays.json");
+	return result;
+}
+
+void CheckPoint::WriteClusters(const ClusterStruct& clusters, boost::gregorian::date date)
+{
+
+	std::string datestr = to_iso_string(date);
+	htri_t exist = H5Lexists(m_file, datestr.c_str(), H5P_DEFAULT);
+	if (exist <= 0) {
+		hid_t temp = H5Gcreate2(m_file, datestr.c_str(), H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+		H5Gclose(temp);
+	}
+	hid_t group = H5Gopen2(m_file, datestr.c_str(), H5P_DEFAULT);
+
+	WriteCluster(clusters.m_households, group);
+	WriteCluster(clusters.m_school_clusters, group);
+	WriteCluster(clusters.m_work_clusters, group);
+	WriteCluster(clusters.m_primary_community, group);
+	WriteCluster(clusters.m_secondary_community, group);
+
+	H5Gclose(group);
+}
+
+void CheckPoint::WriteCluster(const std::vector<Cluster>& clvector, hid_t& group)
+{
+	std::string dsetname = ToString(clvector.front().GetClusterType());
```

Comment:
This line of code will segfault if there are no instances of a type of cluster. Please pass the cluster type to `WriteCluster` as a parameter instead of calling `clvector.front()`.